kong/dns/client.lua

```diff
@@ -0,0 +1,783 @@
+    stats         = {},
+    hosts         = hosts,
+    r_opts        = r_opts,
+    resolv        = opts._resolv or resolv,
+    valid_ttl     = opts.valid_ttl,
+    error_ttl     = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl     = opts.stale_ttl or DEFAULT_STALE_TTL,
+    empty_ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    hosts_cache   = hosts_cache,
+    search_types  = search_types,
+
+    -- TODO: Make the table readonly. But if `string.buffer.encode/decode` and
+    -- `pl.tablex.readonly` are called on it, it will become empty table.
+    --
+    -- quickly accessible constant empty answers
+    EMPTY_ANSWERS = {
+      errcode = EMPTY_RECORD_ERROR_CODE,
+      errstr  = EMPTY_RECORD_ERROR_MESSAGE,
+      ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    },
+  }, MT)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_CODE and self.empty_ttl or self.error_ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = math_min(self.valid_ttl or 0xffffffff, 0xffffffff)  -- 0xffffffff for maximum TTL value
+
+  for _, answer in ipairs(answers) do
+    answer.name = string_lower(answer.name)
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+
+    local answer_type = answer.type
+
+    if answer_type == TYPE_CNAME then
+      cname_answer = answer   -- use the last one as the real cname
+
+    elseif answer_type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer_type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+
+      elseif answer_type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      -- skip the SRV record pointing to itself,
+      -- see https://github.com/Kong/lua-resty-dns-client/pull/3
+      if not (answer_type == TYPE_SRV and answer.target == qname) then
+        table_insert(processed_answers, answer)
+      end
+    end
+  end
+
+  if table_isempty(processed_answers) then
+    if not cname_answer then
+      log(DEBUG, "processed ans:empty")
+      return self.EMPTY_ANSWERS
+    end
+
+    processed_answers[1] = cname_answer
+
+    log(DEBUG, "processed cname:", cname_answer.cname)
+
+  else
+    log(DEBUG, "processed ans:", #processed_answers)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype)
+  local key = name .. ":" .. qtype
+  stats_increment(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local start_time = now()
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
```

Comment:
`options` seems to always be the same and there is probably no concurrency issue? If there is concurrency issue, I would suggest as there is no need for local variable:

```suggestion
  local answers, err = r:query(name, { additional_section = true, qtype = qtype })
```

With no concurrency I would just reuse the `options`, and thus something like this:
```suggestion
  QUERY_OPTS.qtype = qtype
  local answers, err = r:query(name, QUERY_OPTS)
```

Reply:
> With no concurrency issues (which I think there isn't any concurrency issues) I would just reuse the options, and thus something like this:
Yes if there is no concurrency issues, we could write as this. Let me check for it also.

Reply:
I found that there wouldn't be genuine concurrency issues for the `qtype` field in our current scenario. However, in `lua-resty-dns`, `opts.qtype` gets referenced again after calling `sock:receive()`, potentially being altered by other preceding callers who modify `QUERY_OPTS.qtype`. Although checking `qtype` with `TYPE_SOA` (see following code logic) wouldn't lead to concurrency issues because we haven't modified it to `TYPE_SOA` in our DNS client library, I advise against using a static `QUERY_OPTS` in this context. Doing so might introduce potential issues if the underlying `lua-resty-dns` adds logic such as checking `qtype` with `TYPE_A`.

```
r:query()   --- lua-resty-dns: _M.query
    ...
    for i = 1, retrans do
        ...
        ok, err = sock:send(query)
        if not ok then
      ...
        else--
             ...
            for _ = 1, 128 do
                buf, err = sock:receive(4096)      
                                 -- '>>> lua vm yields from here, some one may modify `QUERY_OPTS.qtype`
                ...
                if buf then
                    local answers
                    answers, err = parse_response(buf, id, opts) --- >>> it re-references opts <<<

  ...


parse_response()
 ...
         if opts.qtype == TYPE_SOA then           --- >>> it re-references opts.qtype <<<
            authority_section = true
        end
```

Reply:
I will accept this suggestion, removing the local variable, like the following,

```
  local answers, err = r:query(name, { additional_section = true, qtype = qtype })
```

Reply:
Internal discussion logged here:

After some effort, even though it could work (I have tested my diff as following on using tablepool, it works well.)

```
diff --git a/kong/dns/client.lua b/kong/dns/client.lua
index abc7046f1..45fc4b387 100644
--- a/kong/dns/client.lua
+++ b/kong/dns/client.lua
@@ -2,6 +2,7 @@ local cjson = require("cjson.safe")
 local utils = require("kong.dns.utils")
 local mlcache = require("kong.resty.mlcache")
 local resolver = require("resty.dns.resolver")
+local tablepool = require("tablepool")

 local now = ngx.now
 local log = ngx.log
@@ -21,6 +22,8 @@ local math_min = math.min
 local string_lower = string.lower
 local table_insert = table.insert
 local table_isempty = require("table.isempty")
+local tablepool_fetch = tablepool.fetch
+local tablepool_release = tablepool.release

 local parse_hosts = utils.parse_hosts
 local ipv6_bracket = utils.ipv6_bracket
@@ -33,6 +36,8 @@ local req_dyn_hook_run_hooks = require("kong.dynamic_hook").run_hooks

 -- Constants and default values

+local POOL_DNS_CLIENT = "KONG_DNS_CLIENT"
+
 local DEFAULT_ERROR_TTL = 1     -- unit: second
 local DEFAULT_STALE_TTL = 4
 local DEFAULT_EMPTY_TTL = 30
@@ -407,9 +412,15 @@ local function resolve_query(self, name, qtype)

   local start_time = now()

-  local answers, err = r:query(name, { additional_section = true, qtype = qtype })
+  local options = tablepool_fetch(POOL_DNS_CLIENT, 0, 2)
+  options.additional_section = true
+  options.qtype = qtype
+
+  local answers, err = r:query(name, options)
   r:destroy()

+  tablepool_release(POOL_DNS_CLIENT, options, true)  -- no clear
+
   local query_time = now() - start_time -- the time taken for the DNS query
   local time_str = ("%.3f %.3f"):format(start_time, query_time)
```

1. It makes code hard to read
2. There is another part of  temporary table creation stale_update_task()->self.cache:set(key, { ttl = answers.ttl }, answers), which is also not on hot path. 
3. It is a lua-vm land problem, but currently we have no good method to resolve the table gc problem in lua land
4. The most important is that these table creations are not on hot code path, the hot one of resolved_names table has been resolved by you and me .

So I think we just keep the original logic.